### PR TITLE
Mesh-762 MIP and governance process improvements

### PR DESCRIPTION
**Pull request by vkandy**
Following improvements were made:

- [x] All governance functions required root origin
- [x] Remove `ensure_root` from calls where custom origin is used
- [x] Polymesh committee is managed through group, removed `set_members`
- [x] If committee is composed of a less than 2, referenda are enacted automatically
- [x] If committee is composed of a less than 2, proposals are executed automatically
- [x] A single member of committee can fast track a proposal or create a new referendum
- [x] MIP takes an optional description field
- [x] Voting threshold always assumes "at least" comparision.
- [x] Voting threshold only takes proportion fraction as argument (can be changed through MIP)
- [x] Added Error types for Group and Committee modules

### CODE REVIEW

File: pallets/mips/src/lib.rs [hunk 1]
```diff
@@ -775,30 +808,41 @@ mod tests {
...
-        pub const Four: u64 = 4;
-        pub const Five: u64 = 5;
+    }
+
+    impl group::GroupTrait for Test {
+        fn get_members() -> Vec<IdentityId> {
+            vec![]
+        }
+
+        fn is_member(member_id: &IdentityId) -> bool {
```
Comment (fmiguelgarcia) [inline pallets/mips/src/lib.rs]: Not need it.

File: pallets/mips/src/lib.rs [hunk 2]
```diff
@@ -987,62 +1068,86 @@ mod tests {
...
-            );
+            let alice_acc = AccountKeyring::Alice.public();
+            let (alice_signer, _) = make_account_with_balance(alice_acc, 60).unwrap();
+
+            // Voting majority
+            let root = Origin::system(frame_system::RawOrigin::Root);
 
-            assert_ok!(Mips::emergency_referendum(
-                Origin::signed(1),
+            // assert_err!(
```
Comment (fmiguelgarcia) [inline pallets/mips/src/lib.rs]: commented code: Could we remove it or Could we add a comment?

File: pallets/runtime/src/simple_token.rs [hunk 3]
```diff
@@ -394,6 +394,12 @@ mod tests {
         fn get_members() -> Vec<IdentityId> {
             unimplemented!()
         }
+        fn is_member(_did: &IdentityId) -> bool {
```
Comment (fmiguelgarcia) [inline pallets/runtime/src/simple_token.rs]: Not needed.

File: pallets/runtime/src/simple_token.rs [hunk 4]
```diff
@@ -394,6 +394,12 @@ mod tests {
...
+            unimplemented!()
+        }
+        fn member_count() -> usize {
```
Comment (fmiguelgarcia) [inline pallets/runtime/src/simple_token.rs]: ditto.

File: pallets/runtime/src/voting.rs [hunk 5]
```diff
@@ -582,6 +582,14 @@ mod tests {
         fn get_members() -> Vec<IdentityId> {
             unimplemented!();
         }
+
+        fn is_member(_member_id: &IdentityId) -> bool {
```
Comment (fmiguelgarcia) [inline pallets/runtime/src/voting.rs]: ditto.

File: pallets/runtime/src/voting.rs [hunk 6]
```diff
@@ -582,6 +582,14 @@ mod tests {
...
+            unimplemented!();
+        }
+
+        fn member_count() -> usize {
```
Comment (fmiguelgarcia) [inline pallets/runtime/src/voting.rs]: ditto

File: pallets/staking/src/mock.rs [hunk 7]
```diff
@@ -247,6 +247,10 @@ impl GroupTrait for Test {
     fn is_member(_did: &IdentityId) -> bool {
         true
     }
+
+    fn member_count() -> usize {
```
Comment (fmiguelgarcia) [inline pallets/staking/src/mock.rs]: ditto.

File: pallets/group/src/lib.rs [hunk 8]
```diff
@@ -220,10 +214,14 @@ decl_error! {
...
     fn get_members() -> Vec<IdentityId> {
-        return Self::members();
+        Self::members()
     }
 
     fn is_member(did: &IdentityId) -> bool {
         Self::members().iter().any(|id| id == did)
     }
+
+    fn member_count() -> usize {
```
[the remark this answers is not in the document]
Comment (vkandy) [reply pallets/group/src/lib.rs]: thanks - I thought I removed all references.

File: pallets/mips/src/lib.rs [hunk 9]
```diff
@@ -95,6 +113,8 @@ pub struct MipsMetadata<AcccountId: Parameter, BlockNumber: Parameter, Hash: Par
     proposal_hash: Hash,
     /// The proposal url for proposal discussion.
     url: Option<Url>,
+    /// The proposal description.
+    description: Option<MipDescription>,
```
[the remark this answers is not in the document]
Comment (vkandy) [reply pallets/mips/src/lib.rs]: Perhaps the UI / client can make that call.

File: pallets/mips/src/lib.rs [hunk 10]
```diff
@@ -544,6 +575,11 @@ impl<T: Trait> Module<T> {
             priority.clone(),
             proposal_hash.clone(),
         ));
+
+        // If committee size is too small, enact it.
+        if T::GovernanceCommittee::member_count() < 2 {
+            Self::prepare_to_dispatch(proposal_hash.clone());
```
Comment (adamdossa) [inline pallets/mips/src/lib.rs]: I think this logic needs to be in committee.rs to correctly evalulate the GC quorum when there is only a single GC member. As written here, the GC member won't need to vote at all, and every referenda is automatically passed?

Comment (vkandy) [reply pallets/mips/src/lib.rs]: This is for a special case. Since GC members can create referendums directly, if the committee is composed of a single member, it's enacted immediately. This way single GC member need not create a MIP to enact the referendum they just created.

Without this if block, the process gets a bit tedious for the single member for the same outcome:
1. Member creates a referendum
2. Member creates a MIP proposal to `enact_referendum` with hash from step 1
3. Committee executes `enact_referendum` it because member count is < 2

**Comment by satyamakgec** [comment]
@vkandy isn't it some of the same commits are present in the other PR #254 ?

File: pallets/mips/src/lib.rs (hunk 10 above)
Comment (adamdossa) [reply pallets/mips/src/lib.rs]: Ah OK - that makes sense. Do we not need some logic in committee anyway so that in the case of a normal GC vote a single member can still function? i.e. checking whether the threshold is passed during the proposal function as well in the vote function.

Comment (vkandy) [reply pallets/mips/src/lib.rs]: The original verison had the logic but went missing. I've added the checks in `committee.rs` to execute a proposal when member count < 2. Also added a test case in `committee_test.rs` to ensure proposal is executed when a single member committee makes a proposal.

Comment (vkandy) [reply pallets/mips/src/lib.rs]: @adamdossa there's an edge case to consider: 
1. Say Alice, Bob, Charlie are committee members. 
2. Alice makes a proposal. 
3. Bob and Charlie quit.
4. Alice can't quit since she's the last remaining member. 
The proposal Alice made remains in queue.

**Comment by vkandy** [comment]
> @vkandy isn't it some of the same commits are present in the other PR #254 ? or that PR copies this PR commits ?
This is a branch of an earlier PR which had not been merged at the time of creation of this branch.

File: pallets/mips/src/lib.rs (hunk 10 above)
Comment (adamdossa) [reply pallets/mips/src/lib.rs]: OK. I guess in this case Alice can just make a new proposal, which would now be automatically approved as she is now the only remaining member. If this is the case then I think we can live with this.

The alternative would be to iterate over all proposals every time the GC membership changes and approve any which now meet the new implied threshold - I think this overkill at the moment, provided the first statement above holds.
